sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala

```diff
@@ -54,9 +54,20 @@ trait StringRegexExpression extends Expression
     Pattern.compile(escape(str))
   }
 
+  var lastPatternStr: String = null
+  var compiledPattern: Pattern = null
+
   def nullSafeMatch(input1: Any, input2: Any): Any = {
-    val s = input2.asInstanceOf[UTF8String].toString
-    val regex = if (cache == null) compile(s) else cache
+    val patternStr = input2.asInstanceOf[UTF8String].toString
+    val regex = if (cache == null) {
+      if (!(patternStr).equals(lastPatternStr)) {
```

Comment:
Benefits of this optimization depend on how patterns are distributed in a column. For example, if we have:

| string | pattern|
|-|-|
|s1|pattern1|
|s2|pattern2|
|s3|pattern1|
|s4|pattern2|

in that case, this optimization can lead to slowdown.